docs: Adding a push to deploy on github guide (#211)
shorebirdtech/docs: docs/guides/push_to_deploy_on_github.md

---
sidebar_position: 8
title: 🫸 Push to Deploy on GitHub

> description: Automatically deploy to Shorebird by pushing to your GitHub repository.

Reviewer: ```suggestion
description: Configure automated Shorebird deployments via GitHub.
```

---

# 🫸 Push to Deploy on GitHub

:::info

> If not yet, we recommend reading the base [Github Integration](/ci/github) instructions before dwelling into this guide.

Reviewer: ```suggestion
We recommend reading the [Github Integration](/ci/github) instructions before diving into this guide.
```

is intentionally simple so that it can be easily adapted to different requirements and contexts.

## 🥅 Goal

We will be implementing an automation in a project that have the following goals:

- Any branches spinned off from `main` that starts with the prefix `release/` will trigger a
  release at Shorebird.
- Any additional commits on those branches will trigger patches.

## 📝 Prerequisites

Before anything, if you don't have a Shorebird account, be sure to create one at
[shorebird.dev](https://shorebird.dev) (Don't worry, it is free to start have a generous quota
that is way more than enough for this guide).

You will also need a GitHub account and a project repository created. If you don't have an account
yet, create one at [GitHub](https://github.com), and create a new repository, which for this guide
we will call `shorebird_push_to_deploy`.

## 🚀 Getting Started

First thing, lets create a new Flutter project simply by running:

```bash
flutter create shorebird_push_to_deploy
```

Be sure to initialize your project as a git repository:

```bash
cd shorebird_push_to_deploy
git init
git remote add origin <THE_URL_OF_YOUR_REPOSITORY>
```

> Next we need to initialize shorebird in it! If you have doubts on how to do so, check the
> [Code Push Getting Started](/code_push/initialize) guide.

Reviewer: ```suggestion
Next we need to initialize shorebird in it! Check out the [Code Push Getting Started](/code_push/initialize) guide for more info.
```

## 🛠️ Setting up GitHub Actions

We provide a collection of official Github Actions to help you with the integration. For this guide

> we will be using `shorebird-setup`, `shorebird-release` and `shorebird-patch`, you can check more
> details about them at our [Github Integration Documentation](ci/github).

Reviewer: ```suggestion
we will be using `shorebird-setup`, `shorebird-release` and `shorebird-patch`, you can learn more
about them at our [Github Integration Documentation](ci/github).
```

Create a new file at `.github/workflows/push_to_deploy_android.yml` with the following content:

```yaml
name: Shorebird Android

on:
  push:
    branches:
      - releases/*

env:
  SHOREBIRD_TOKEN: ${{ secrets.SHOREBIRD_TOKEN }}

jobs:
  push_to_deploy_android:
    defaults:
      run:
        shell: bash

    runs-on: ubuntu-latest

    steps:
      - name: 📚 Git Checkout
        uses: actions/checkout@v3

      - name: 🐦 Setup Shorebird
        uses: shorebirdtech/setup-shorebird@v1

      - name: 🚀 Shorebird Release
        if: ${{ github.event.created }}
        uses: shorebirdtech/shorebird-release@v0
        with:
          platform: android

      - name: 🚀 Shorebird Patch
        if: ${{ !github.event.created }}
        uses: shorebirdtech/shorebird-patch@v0
        with:
          platform: android
```

> The content of this file is pretty straightforward. It defines a workflow that will be triggered
> whenever a push is made to a branch that starts with `releases/`.

Reviewer: ```suggestion
Let's take a closer look at the above workflow.
The workflow will be triggered whenever a push is made to a branch that starts with `releases/`.
```

It will then run the clone of the repository and setup the Shorebird CLI.

Next the workflow is divided into two conditional steps. If the push is the one that created the
branch, it will trigger a release. Otherwise it will trigger a patch.

> Note that the `SHOREBIRD_TOKEN` is being used as an environment variable. This is a secret that
> should be stored in your repository settings. You might have done that already if you checked the
> [GitHub Integration Documentation](/ci/github) mentioned above, but if not be sure to check it
> and configure the authentication in the action for your Shorebird account.

Reviewer: ```suggestion
:::note
The `SHOREBIRD_TOKEN` is used as an environment variable. This is a secret that
should be stored in your repository settings. If you have not already configured the secrets, check out the
[GitHub Integration Documentation](/ci/github) mentioned above.
:::
```

> If all went right, you should be able to push the changes to your repository and see the workflow
> running at the Actions tab of your repository.

Reviewer: ```suggestion
That's it! You should be able to push the changes to your repository and see the workflow
running in the actions tab of your repository.
```

> Give it a try! Create a new branch called `releases/1.0.0` and push it to your repository. Then

Reviewer: Maybe we should include the git commands for completeness here?

Author: Added, lmk if that is what you had in mind

> once the workflow finishes, check the Shorebird dashboard and you should see a new release there.

Reviewer: ```suggestion
once the workflow finishes, check the [Shorebird console](https://console.shorebird.dev) and you should see a new release for your app.
```

> Next you can try pushing a new commit to the same branch and see a patch being created 🎉.

Reviewer: Maybe we should also include an example of a small code change (like changing the color similar to the quick start guide) and also include the git commands used.

Author: Added, lmk if that is what you had in mind

> ## 🎬 End Notes

Reviewer: ```suggestion
## 🎬 Summary
```

> Like mentined in the beginning, what we presented here is just one way of setting up a Push to Deploy

Author: ```suggestion
Like mentioned in the beginning, what we presented here is just one way of setting up a Push to Deploy
```

> Like mentioned in the beginning, what we presented here is just one way of setting up a Push to Deploy
> ideally the reader should adapt it to their own needs and context.

Reviewer: ```suggestion
As mentioned, the workflow presented here is just one way of setting up an automated push to deploy workflow.
Feel free to adapt it based on your teams needs and existing processes.
```

> And it can be expanded in many ways. For example:

Reviewer: ```suggestion
Some ways the workflow can be expanded are:
```

> And what was presented here could also be expanded in many ways. For example:

Author: ```suggestion
And it can be expanded in many ways. For example:
```

- Instead of directly commiting to the branch after the release was made. Developers would land their
  changes and fixes on their main branch, and then `cherry-pick` it to the release branch!
- Both `shorebird-release` and `shorebird-patch` returns the version/patch-number created. The workflow
  could be expanded in order for tags to be created in the repository, using the version number returned
  allowing snapshots of the code to be easily identified.

> - etc.

Reviewer: ```suggestion
```

> What was explained here is implemented in our [Time Shift App](https://github.com/shorebirdtech/time_shift).

Reviewer: ```suggestion
See this workflow in action by checking out our [Time Shift App](https://github.com/shorebirdtech/time_shift).
```

> If you want see an actual implementation, feel free to check it out!

Reviewer: ```suggestion
```
